Sources/SourceKitLSP/Swift/CodeActions/ApplyDeMorganLaw.swift

```diff
@@ -0,0 +1,730 @@
+    ///  1. The negated expression is a composite expression, and
+    ///  2. The original expression is not a single.
+    ///
+    /// - Note: A composite expression is one of `AsExprSyntax`, `AwaitExprSyntax`, `InfixOperatorExprSyntax`,
+    /// `IsExprSyntax`, or `TryExprSyntax`.
+    ///
+    /// An example of mapping is as follows,
+    /// ```swift
+    /// // original expression
+    /// ((!a || b is String))
+    /// // expression of interest
+    /// !a || b is String
+    /// !a /* negated to and mapped to */ a
+    /// b is String /* negated to */ Negation(b is String) /* mapped to */ !(b is String)
+    /// // negated expression of interest
+    /// a && !(b is String)
+    /// // mapped to
+    /// !((a && !(b is String)))
+    /// ```
+    func map(negation: Negation) -> Negated {
+      let negatedExpr =
+        switch self {
+        case .single(let tuple, _):
+          PrefixOperatorExprSyntax(
+            operator: negation.exprType.negationPrefix,
+            expression: tuple.map(complementExpr: negation.expr)
+          )
+        case .other:
+          PrefixOperatorExprSyntax(
+            operator: negation.exprType.negationPrefix,
+            expression: negation.expr.isComposite
+              ? ExprSyntax(TupleExprSyntax(onlyExpr: negation.expr)) : negation.expr
+          )
+        }
+      return (ExprSyntax(negatedExpr), .negation)
+    }
+
+    /// Substitutes the expression of interest with the negated expression and returns the result.
+    func map(negatedExpr: some ExprSyntaxProtocol) -> ExprSyntax {
+      switch self {
+      case .single(let tuple, _):
+        ExprSyntax(tuple.map(complementExpr: ExprSyntax(negatedExpr)))
+      case .other:
+        ExprSyntax(negatedExpr)
+      }
+    }
+  }
+
+  /// Represents a negation expression with `expr` being its only sub-expression.
+  private struct Negation {
+    var exprType: ExprType
+    var expr: ExprSyntax
+
+    init(exprType: ExprType, expr: some ExprSyntaxProtocol) {
+      self.exprType = exprType
+      self.expr = ExprSyntax(expr)
+    }
+
+    /// If `exprType` is `nil`, this indicates the negation expression is at the root-level and `exprType` will be
+    /// inferred from the prefix of `prefixExpr`. Otherwise `prefixExpr` will be checked for its prefix's agreement
+    /// with the given `exprType`.
+    init?(_ prefixExpr: PrefixOperatorExprSyntax, exprType: ExprType? = nil) {
+      guard let localExprType = ExprType(operator: prefixExpr.operator) else {
+        return nil
+      }
+      if let exprType, exprType != localExprType {
+        return nil
+      }
+      self.init(
+        exprType: localExprType,
+        expr: prefixExpr.expression.with(
+          \.leadingTrivia,
+          prefixExpr.leadingTrivia + prefixExpr.expression.leadingTrivia
+        )
+      )
+    }
+  }
+
+  /// Represents the negated expression of an instance of propositions.
+  private struct NegatedPropositions {
+    struct Operator {
```

Comment:
I feel like the `Operator` type is a little overkill here. Wouldn’t it easier to just have a function that returns the negated operator’s `TokenKind` and `ExprType`?

Reply:
We need something more than `TokenKind` here, (i.e. to tell AND or OR)

Reply:
Do we need a whole type for that though? Wouldn’t it be sufficient to have

```swift
// In `NegatedPropositions`
var exprType: ExprType? {
  switch expr.operator.tokenKind {
  case .binaryOperator("|"):
    return .bitwise
  case .binaryOperator("&"):
    return .bitwise
  case .binaryOperator("||"):
    return .boolean
  case .binaryOperator("&&"):
    return .boolean
  default:
    return nil
  }
}
```

And where we check for `operator.kind == .and` also just check `expr.operator.tokenKind == .binaryOperator("&") || expr.operator.tokenKind == .binaryOperator("&&")`. 

Or alternatively, if you prefer add

```swift
extension TokenKind {
  var exprType: ExprType? {
    switch expr.operator.tokenKind {
    case .binaryOperator("|"):
      return .bitwise
    case .binaryOperator("&"):
      return .bitwise
    case .binaryOperator("||"):
      return .boolean
    case .binaryOperator("&&"):
      return .boolean
    default:
      return nil
    }
  }

  var isAnd: Bool {
    switch expr.operator.tokenKind {
    case .binaryOperator("&"):
      return .bitwise
    case .binaryOperator("&&"):
      return .boolean
    default:
      return nil
    }
  }
}
```

The set of `and`/`or` operators hasn’t changed since C was introduced in the 70s (or even before in other language), so I don’t think we need an abstraction layer for it.